### PR TITLE
Merging initial build to change error reporting

### PR DESCRIPTION
**Pull request by AnmolMajithia**
The current breaker is set to " | ", will be updated when Sasa or someone gives me any inputs.

The bulky error message sample replies function will be ported to an external file or some variant of that once Amar has some suggestions for that.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -66,7 +66,7 @@ superset/app/
 *.js.map
 node_modules
 npm-debug.log*
-superset/static/assets
+# superset/static/assets
```
Comment (AnmolMajithia) [inline .gitignore:69]: Delete line

File: superset/db_engine_specs/base.py [hunk 2]
```diff
@@ -72,6 +72,8 @@
     from superset.connectors.sqla.models import TableColumn
     from superset.models.core import Database
 
+import requests
```
Comment (AnmolMajithia) [inline superset/db_engine_specs/base.py:75]: remove

File: superset/db_engine_specs/base.py [hunk 3]
```diff
@@ -751,6 +753,11 @@ def _extract_error_message(cls, ex: Exception) -> str:
     def extract_errors(
         cls, ex: Exception, context: Optional[Dict[str, Any]] = None
     ) -> List[SupersetError]:
+        ### IKIGAI
+        # Personal Variables
+        DB_NAME = os.environ.get("DB_NAME")
```
Comment (AnmolMajithia) [inline superset/db_engine_specs/base.py:758]: Where to move

File: superset/db_engine_specs/base.py [hunk 4]
```diff
@@ -768,12 +775,14 @@ def extract_errors(
                     )
                 ]
 
+        # Moved API call to superset/utils/core.py to catch broader set of errors
```
Comment (AnmolMajithia) [inline superset/db_engine_specs/base.py:778]: Move to a more logical place

File: superset/utils/core.py [hunk 5]
```diff
@@ -116,6 +116,8 @@
     from superset.connectors.base.models import BaseColumn, BaseDatasource
     from superset.models.core import Database
 
+#Custom Imports (ikigai):
```
Comment (AnmolMajithia) [inline superset/utils/core.py:119]: remove

File: superset/utils/core.py [hunk 6]
```diff
@@ -116,6 +116,8 @@
...
+import requests
```
Comment (AnmolMajithia) [inline superset/utils/core.py:120]: move to top

File: superset/utils/core.py [hunk 7]
```diff
@@ -617,8 +619,81 @@ def json_int_dttm_ser(obj: Any) -> float:
...
+            " If it still exists, contact IkigaiLabs for assistance.",
+        "RESOURCE TIMEOUT": "Your access to this resource has timed out, try re running the query." \
+            " If the problem still persists try relogging in or contact IkigaiLabs.",
+        "RETRY ATTEMPT ERROR": "We tried to run your query again and again, still looks like we are running in an error." \
+            "How about contacting IkigaiLabs for assistance?",
+        "REFRESH DATASET ERROR": "Looks like we had an Internal Error while Refreshing the Dataset." \
+            "Try re running the query or contact IkigaiLabs for further assisstance.",
+        "PDFS RETRIABLE ERROR": "This is a rare incident. Re running the query might solve it! :)"
+    }
+    return error_messages.get(error, "Looks like we ran into an unknown error. Please contact IkigaiLabs for assistance.")
```
Comment (AnmolMajithia) [inline superset/utils/core.py:671]: ikigailabs --> support

File: superset/utils/core.py [hunk 8]
```diff
@@ -617,8 +619,81 @@ def json_int_dttm_ser(obj: Any) -> float:
...
+
+def error_message_beautifier_dremio(response):
+    project = response.json()['project']['name'].strip()
+    dataset = response.json()['dataset']['name'].strip()
+    chart = response.json()['chart']['name'].strip()
+    error = response.json()['error_type'].strip()
+    breaker = " |"
```
Comment (AnmolMajithia) [inline superset/utils/core.py:678]: move this to the plave we move the dict too

File: superset/utils/core.py [hunk 9]
```diff
@@ -617,8 +619,81 @@ def json_int_dttm_ser(obj: Any) -> float:
...
+    if len(chart) != 0:
+        msg.append(f'{breaker} Chart: {chart} ')
+    if len(error) != 0:
+        msg.append(f'{breaker} Error Type: {error} ')
+        msg.append( f'{breaker} Suggestion: {error_type_suggestor(error)}')
+    return "".join(msg)
+    
 
 def error_msg_from_exception(ex: Exception) -> str:
+    DB_NAME = os.environ.get("DB_NAME")
```
Comment (AnmolMajithia) [inline superset/utils/core.py:694]: relocate??

File: superset/utils/core.py [hunk 10]
```diff
@@ -638,11 +713,37 @@ def error_msg_from_exception(ex: Exception) -> str:
...
-    return msg or str(ex)
+    
+    if msg:
+        return msg
+    
+    if DB_NAME not in str(ex):
+        return str(ex)
+    
+    error_msg = str(ex)
+    if "ikigai-datasets-dev" in error_msg:
```
Comment (AnmolMajithia) [inline superset/utils/core.py:724]: remove

File: superset/utils/core.py [hunk 11]
```diff
@@ -638,11 +713,37 @@ def error_msg_from_exception(ex: Exception) -> str:
...
+        error_msg = error_msg.replace("ikigai-datasets-dev", "ikigai-datasets-virtual")
+    
+    PARAMS = {
```
Comment (AnmolMajithia) [inline superset/utils/core.py:727]: PARAMS --> PAYLOAD

File: superset/utils/core.py [hunk 12]
```diff
@@ -638,11 +713,37 @@ def error_msg_from_exception(ex: Exception) -> str:
...
+        'error_string':error_msg,
```
Comment (AnmolMajithia) [inline superset/utils/core.py:728]: error_msg --> str(ex)

File: superset/utils/core.py [hunk 13]
```diff
@@ -638,11 +713,37 @@ def error_msg_from_exception(ex: Exception) -> str:
...
+    }
+
+    # Send to API:
+    URL = BASE_URL+DREMIO_PARSE_ENDPOINT
```
Comment (AnmolMajithia) [inline superset/utils/core.py:732]: urljoin

File: superset/utils/core.py [hunk 14]
```diff
@@ -638,11 +713,37 @@ def error_msg_from_exception(ex: Exception) -> str:
...
+    response = requests.post(
+        url=URL, 
+        data=json.dumps(PARAMS))
+
+    # If API does not respond as expected:
+    if response.status_code != 200:
+        return f'[{response.status_code}] Unable to connect to Error Reporting API please contact IkigaiLabs for further assistance.'
```
Comment (AnmolMajithia) [inline superset/utils/core.py:739]: support

File: superset/utils/core.py [hunk 15]
```diff
@@ -638,11 +713,37 @@ def error_msg_from_exception(ex: Exception) -> str:
...
+    
+    return error_message_beautifier_dremio(response)
 
 
 def markdown(raw: str, markup_wrap: Optional[bool] = False) -> str:
     safe_markdown_tags = [
+
```
Comment (AnmolMajithia) [inline superset/utils/core.py:746]: remove

File: superset/db_engine_specs/base.py [hunk 16]
```diff
@@ -751,6 +752,11 @@ def _extract_error_message(cls, ex: Exception) -> str:
     def extract_errors(
         cls, ex: Exception, context: Optional[Dict[str, Any]] = None
     ) -> List[SupersetError]:
+        ### IKIGAI
+        # Personal Variables
+        DB_NAME = os.environ.get("DB_NAME")
+        DB_NAME_REPLACE = os.environ.get("DB_NAME_REPLACE")
```
Comment (AnmolMajithia) [inline superset/db_engine_specs/base.py:758]: ALTERNATIVE_DB_NAME

File: superset/utils/core.py [hunk 17]
```diff
@@ -617,8 +618,81 @@ def json_int_dttm_ser(obj: Any) -> float:
 def json_dumps_w_dates(payload: Dict[Any, Any]) -> str:
     return json.dumps(payload, default=json_int_dttm_ser)
```
Comment (simjay) [inline superset/utils/core.py:620]: eva green the dreamers

File: superset/utils/core.py [hunk 18]
```diff
@@ -617,8 +618,81 @@ def json_int_dttm_ser(obj: Any) -> float:
...
 
+def error_type_suggestor(error): # TODO: port this into maybe a file? ask amar for suggestions
```
Comment (simjay) [inline superset/utils/core.py:621]: def get_error_description

Comment (simjay) [reply superset/utils/core.py:621]: and move to the ikigai_utils.py

File: superset/utils/core.py [hunk 19]
```diff
@@ -617,8 +618,81 @@ def json_int_dttm_ser(obj: Any) -> float:
...
+            " If the problem still persists try relogging in or contact support.",
+        "RETRY ATTEMPT ERROR": "We tried to run your query again and again, still looks like we are running in an error." \
+            "How about contacting support for assistance?",
+        "REFRESH DATASET ERROR": "Looks like we had an Internal Error while Refreshing the Dataset." \
+            "Try re running the query or contact support for further assisstance.",
+        "PDFS RETRIABLE ERROR": "This is a rare incident. Re running the query might solve it! :)"
+    }
+    return error_messages.get(error, "Looks like we ran into an unknown error. Please contact support for assistance.")
+
+def error_message_beautifier_dremio(response):
```
Comment (simjay) [inline superset/utils/core.py:672]: lets move to ikigai_utils.py

rename to def parse_error_components

File: superset/utils/core.py [hunk 20]
```diff
@@ -619,6 +622,17 @@ def json_dumps_w_dates(payload: Dict[Any, Any]) -> str:
 
 
 def error_msg_from_exception(ex: Exception) -> str:
+    '''Superset function [DONT CHANGE NAME] with minor changes to assist with Dremio Error Reporting.
```
Comment (AnmolMajithia) [inline superset/utils/core.py:625]: Remove "Dont change name"

File: superset/utils/core.py [hunk 21]
```diff
@@ -619,6 +622,17 @@ def json_dumps_w_dates(payload: Dict[Any, Any]) -> str:
...
+    Main Use:
```
Comment (AnmolMajithia) [inline superset/utils/core.py:626]: remove

File: superset/utils/core.py [hunk 22]
```diff
@@ -619,6 +622,17 @@ def json_dumps_w_dates(payload: Dict[Any, Any]) -> str:
 
 
 def error_msg_from_exception(ex: Exception) -> str:
+    '''Superset function [DONT CHANGE NAME] with minor changes to assist with Dremio Error Reporting.
+    Main Use: 
+    - Captures error message
+    - Checks if error message has "Dremio" in it
+        - If not does not affect supersets pipeline
+        - If yes stores it in "PAYLOAD"
+    - Call API and send response to parser
```
Comment (AnmolMajithia) [inline superset/utils/core.py:631]: explain api

Comment (AnmolMajithia) [reply superset/utils/core.py:631]: send response to parse-dremio-error

File: superset/utils/core.py [hunk 23]
```diff
@@ -638,7 +652,28 @@ def error_msg_from_exception(ex: Exception) -> str:
...
+    
+    if DB_NAME not in str(ex):
+        return str(ex)
+    
+    PAYLOAD = {
+        'error_string':str(ex),
+    }
+
+    # Send to API:
+    URL = urljoin(BASE_URL, DREMIO_PARSE_ENDPOINT)
```
Comment (AnmolMajithia) [inline superset/utils/core.py:667]: (suggestion) - move api call to custom file?

File: superset/utils/core.py [hunk 24]
```diff
@@ -638,7 +652,28 @@ def error_msg_from_exception(ex: Exception) -> str:
...
         elif ex.message:  # type: ignore
             msg = ex.message  # type: ignore
-    return msg or str(ex)
+    
+    if msg:
+        return msg
+    
+    if DB_NAME not in str(ex):
+        return str(ex)
+
```
Comment (AnmolMajithia) [inline superset/utils/core.py:661]: comment

File: superset/utils/core.py [hunk 25]
```diff
@@ -638,7 +652,28 @@ def error_msg_from_exception(ex: Exception) -> str:
             msg = ex.message.get("message")  # type: ignore
         elif ex.message:  # type: ignore
             msg = ex.message  # type: ignore
-    return msg or str(ex)
+    
+    if msg:
+        return msg
+
```
Comment (AnmolMajithia) [inline superset/utils/core.py:658]: comment

File: superset/utils/core.py [hunk 26]
```diff
@@ -638,7 +652,28 @@ def error_msg_from_exception(ex: Exception) -> str:
...
             msg = ex.message  # type: ignore
-    return msg or str(ex)
+    
+    if msg:
+        return msg
+    
+    if DB_NAME not in str(ex):
+        return str(ex)
+    
+    PAYLOAD = {
```
Comment (AnmolMajithia) [inline superset/utils/core.py:662]: remove capitalize

File: superset/utils/core.py (hunk 23 above)
Comment (AnmolMajithia) [inline superset/utils/core.py:667]: URL - remove caps

File: superset/utils/ikigai_utils.py [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+    return error_messages.get(error, "Looks like we ran into an unknown error. Please contact support for assistance.")
+
+
+def parse_error_components(response):
+    '''Fetches components from APIs response and generates a helpful string.'''
+    project = response.json()['project']['name'].strip()
+    dataset = response.json()['dataset']['name'].strip()
+    chart = response.json()['chart']['name'].strip()
+    error = response.json()['error_type'].strip()
+    breaker = " |" # TODO: move to place contents of above function will be moved to
```
Comment (AnmolMajithia) [inline superset/utils/ikigai_utils.py:66]: remove TODO

File: superset/utils/core.py [hunk 28]
```diff
@@ -619,6 +622,16 @@ def json_dumps_w_dates(payload: Dict[Any, Any]) -> str:
 
 
 def error_msg_from_exception(ex: Exception) -> str:
+    '''Superset function with minor changes to assist with Dremio Error Reporting.
```
Comment (simjay) [inline superset/utils/core.py:625]: Use """ for docstring, not '''

File: superset/utils/ikigai_utils.py [hunk 29]
```diff
@@ -0,0 +1,81 @@
+'''
+All custom functions implemented here
+If there is custom code elsewhere inside a function, 
+the function is from superset so best not to mess with it much :)
+'''
+
+def get_error_description(error):
+    '''Dictionary to return user friendly error description based on error type.'''
```
Comment (simjay) [inline superset/utils/ikigai_utils.py:8]: use """ for docstring, instead of '''

**Pull request closed without merging**